examples/ini.rs

```diff
@@ -0,0 +1,292 @@
+//! A basic rather tolerant ini parser, which most certainly does not obey to the standard
+use dangerous::{Error, Expected, Reader};
+use std::io::{self, Read};
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let mut input_data = Vec::new();
+    io::stdin().read_to_end(&mut input_data)?;
+    let input = dangerous::input(input_data.as_ref());
+    match input.read_all::<_, _, Expected>(read_ini) {
+        Ok(ini) => println!("{:#?}", ini),
+        Err(e) => eprintln!("{:#}", e),
+    };
+    Ok(())
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct Pair<'a> {
+    name: &'a str,
+    value: &'a str,
+}
+
+#[derive(Default, Debug, PartialEq, Eq)]
+struct Document<'a> {
+    globals: Vec<Pair<'a>>,
+    sections: Vec<Section<'a>>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct Section<'a> {
+    name: &'a str,
+    properties: Vec<Pair<'a>>,
+}
+
+fn read_ini<'i, E>(r: &mut Reader<'i, E>) -> Result<Document<'i>, E>
+where
+    E: Error<'i>,
+{
+    skip_whitespace_or_comment(r, ConsumeTo::NextToken);
+    if r.at_end() {
+        return Ok(Document::default());
+    }
+    let (globals, sections) = match r.peek_u8()? {
+        b'[' => (vec![], read_sections(r)?),
+        _ => (
+            read_zero_or_more_properties_until_section(r)?,
+            read_sections(r)?,
+        ),
+    };
+    Ok(Document { globals, sections })
+}
+
+fn read_sections<'i, E>(r: &mut Reader<'i, E>) -> Result<Vec<Section<'i>>, E>
+where
+    E: Error<'i>,
+{
+    let mut sections = Vec::new();
+    while !r.at_end() {
+        sections.push(read_section(r)?);
+    }
+    Ok(sections)
+}
+
+fn read_zero_or_more_properties_until_section<'i, E>(
+    r: &mut Reader<'i, E>,
+) -> Result<Vec<Pair<'i>>, E>
+where
+    E: Error<'i>,
+{
+    let mut out = Vec::new();
+    fn is_bare_text(c: u8) -> bool {
+        !c.is_ascii_whitespace() && c != b'=' && c != b'\n' && c != b'['
+    }
+
+    skip_whitespace_or_comment(r, ConsumeTo::NextToken);
+    while !(r.at_end() || matches!(r.peek_u8(), Ok(b'['))) {
```

Comment:
Rather than using `matches!`, I would use `Reader::peek_eq(..)` like so `!(r.at_end() || r.peek_eq(b"["))`

Reply:
Much better!
I wonder if peeking with a slice of one byte can be optimized to just be peeking with a single byte, or if a method that handles only a byte would be beneficial here.

Applied!